### PR TITLE
#21900: Fallback to local DNS when no other nameservers are known (2)

### PR DESCRIPTION
**Pull request by rl1987**
https://trac.torproject.org/projects/tor/ticket/21900

### CODE REVIEW

File: src/feature/relay/dns.c [hunk 1]
```diff
@@ -1357,6 +1357,36 @@ evdns_err_is_transient(int err)
...
+configured_nameserver_address(const size_t idx)
+{
+ struct sockaddr_storage sa;
+ ev_socklen_t sa_len = sizeof(sa);
+
+ if (evdns_base_get_nameserver_addr(the_evdns_base, (int)idx,
+                                    (struct sockaddr *)&sa,
+                                    sa_len) > 0) {
+   tor_addr_t *tor_addr = tor_malloc(sizeof(tor_addr_t));
+   tor_addr_from_sockaddr(tor_addr, (const struct sockaddr *)&sa, NULL);
```
Comment (nmathewson) [inline src/feature/relay/dns.c]: This call is allowed to fail; we need to check its return value.

File: src/feature/relay/dns.c [hunk 2]
```diff
@@ -1357,6 +1357,36 @@ evdns_err_is_transient(int err)
...
+ * Return address of configured nameserver in <b>the_evdns_base</b>
+ * at index <b>idx</b>.
+ */
+tor_addr_t *
+configured_nameserver_address(const size_t idx)
+{
+ struct sockaddr_storage sa;
+ ev_socklen_t sa_len = sizeof(sa);
+
+ if (evdns_base_get_nameserver_addr(the_evdns_base, (int)idx,
```
Comment (nmathewson) [inline src/feature/relay/dns.c]: This function wasn't added to Libevent until 2.1.6-beta; We only require 2.0.x, I think.  If you want to use it, we need to check for its presence in autoconf, and only call it when it's available.

Comment (nmathewson) [reply src/feature/relay/dns.c]: Oh wait, I see that you added a check later.  Great.

File: src/feature/relay/dns.c (hunk 1 above)
Comment (rl1987) [reply src/feature/relay/dns.c]: Fixed in b7edfcbf6bb3a27d914ad883ae75413a4d25739a.

**Pull request merged**